File: v2/golang/client/set_score_test.go
package client

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

// TestUpdateRefusesScoreInsteadOfDroppingIt fixa a falha medida em 2026-08-15.
//
// PATCH /api/v1/records/{id} não tem campo score: o servidor responde 200 e
// descarta a chave. Update com {"score": 8} devolvia sucesso e não gravava
// nada — a mesma forma do defeito do campo `archived` em 2026-06-16.
//
// O servidor de teste registra se foi chamado: a guarda tem que barrar ANTES
// do transporte, senão o cliente ainda gastaria uma requisição para nada.
func TestUpdateRefusesScoreInsteadOfDroppingIt(t *testing.T) {
	serverWasCalled := false
	server := httptest.NewServer(http.HandlerFunc(func(responseWriter http.ResponseWriter, request *http.Request) {
		serverWasCalled = true
		responseWriter.WriteHeader(http.StatusOK)
	}))
	defer server.Close()
	memory := NewMemory("test-key", WithURL(server.URL))

	err := memory.Update(context.Background(), 42, map[string]interface{}{"score": 8})

	if err == nil {
		t.Fatal("Update aceitou score — antes desta guarda isso devolvia sucesso e não gravava nada")
	}
	if !strings.Contains(err.Error(), "SetScore") {
		t.Errorf("a mensagem tem que nomear o método que funciona, got: %v", err)
	}
	if serverWasCalled {
		t.Error("a guarda deixou a requisição sair — tem que barrar antes do transporte")
	}
}

// TestUpdateRefusesScoreAlongsideOtherFields cobre o caso perigoso: o summary
// seria gravado e o score não, produzindo sucesso PARCIAL — perda silenciosa
// por outra porta.
func TestUpdateRefusesScoreAlongsideOtherFields(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(responseWriter http.ResponseWriter, _ *http.Request) {
		responseWriter.WriteHeader(http.StatusOK)
	}))
	defer server.Close()
	memory := NewMemory("test-key", WithURL(server.URL))

	err := memory.Update(context.Background(), 42, map[string]interface{}{
		"summary": "novo resumo",
		"score":   8,
	})

	if err == nil {
		t.Fatal("Update aceitou score junto de outro campo — sucesso parcial é perda silenciosa")
	}
}

// TestUpdateStillAcceptsOrdinaryFields: a guarda não pode desligar o Update
// para todo mundo.
func TestUpdateStillAcceptsOrdinaryFields(t *testing.T) {
	var capturedPath, capturedMethod string
	server := httptest.NewServer(http.HandlerFunc(func(responseWriter http.ResponseWriter, request *http.Request) {
		capturedPath, capturedMethod = request.URL.Path, request.Method
		responseWriter.WriteHeader(http.StatusOK)
	}))
	defer server.Close()
	memory := NewMemory("test-key", WithURL(server.URL))

	if err := memory.Update(context.Background(), 42, map[string]interface{}{"summary": "novo"}); err != nil {
		t.Fatalf("Update de campo comum falhou: %v", err)
	}
	if capturedMethod != http.MethodPatch || capturedPath != "/api/v1/records/42" {
		t.Errorf("rota errada: %s %s", capturedMethod, capturedPath)
	}
}

// TestSetScoreHitsTheDurableRoute prova ROTA e CORPO — é o que separa esta
// correção de repetir o defeito num endereço novo.
func TestSetScoreHitsTheDurableRoute(t *testing.T) {
	var capturedPath, capturedMethod string
	var capturedBody map[string]interface{}
	server := httptest.NewServer(http.HandlerFunc(func(responseWriter http.ResponseWriter, request *http.Request) {
		capturedPath, capturedMethod = request.URL.Path, request.Method
		rawBody, _ := io.ReadAll(request.Body)
		_ = json.Unmarshal(rawBody, &capturedBody)
		responseWriter.Header().Set("Content-Type", "application/json")
		_, _ = responseWriter.Write([]byte(`{"status":"ok"}`))
	}))
	defer server.Close()
	memory := NewMemory("test-key", WithURL(server.URL))

	if err := memory.SetScore(context.Background(), 42, 8); err != nil {
		t.Fatalf("SetScore falhou: %v", err)
	}

	if capturedMethod != http.MethodPost {
		t.Errorf("método = %s, esperado POST", capturedMethod)
	}
	if capturedPath != "/api/v1/records/set-score" {
		t.Errorf("rota = %s, esperado /api/v1/records/set-score (o comando replicado)", capturedPath)
	}
	if capturedBody["score"] != float64(8) {
		t.Errorf("corpo sem score correto: %v", capturedBody)
	}
	ids, hasIDs := capturedBody["ids"].([]interface{})
	if !hasIDs || len(ids) != 1 || ids[0] != float64(42) {
		t.Errorf("corpo sem ids=[42]: %v", capturedBody)
	}
}

// TestSetScoreValidatesRange: a faixa do schema é 1..10, pontas inclusive.
func TestSetScoreValidatesRange(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(responseWriter http.ResponseWriter, _ *http.Request) {
		responseWriter.Header().Set("Content-Type", "application/json")
		_, _ = responseWriter.Write([]byte(`{"status":"ok"}`))
	}))
	defer server.Close()
	memory := NewMemory("test-key", WithURL(server.URL))

	for _, invalidScore := range []int{0, -1, 11, 99} {
		if err := memory.SetScore(context.Background(), 42, invalidScore); err == nil {
			t.Errorf("SetScore aceitou %d, fora de [1,10]", invalidScore)
		}
	}
	// As pontas são válidas: a guarda não pode encolher a faixa do schema.
	for _, validScore := range []int{1, 10} {
		if err := memory.SetScore(context.Background(), 42, validScore); err != nil {
			t.Errorf("SetScore rejeitou %d, que está dentro da faixa: %v", validScore, err)
		}
	}
}

// TestSetScoreRequiresConnection espelha o contrato dos outros métodos.
func TestSetScoreRequiresConnection(t *testing.T) {
	memory := &Memory{}

	if err := memory.SetScore(context.Background(), 42, 8); err != ErrEmptyAPIKey {
		t.Errorf("SetScore sem conexão = %v, esperado ErrEmptyAPIKey", err)
	}
}

File: v2/typescript/src/test/set_score.test.ts
/**
 * O score não pode ser alterado por update() — falha medida em 2026-08-15.
 *
 * PATCH /api/v1/records/{id} não tem campo score: o servidor responde 200 e
 * descarta a chave. Um cliente que chamasse update(id, { score: 8 }) recebia
 * sucesso e nada era gravado — a mesma forma do defeito do campo `archived`.
 *
 * A correção não é dividir a chamada em duas por baixo dos panos: isso criaria
 * sucesso PARCIAL (summary grava, score falha), que é a mesma perda silenciosa
 * por outra porta. update() lança e nomeia setScore().
 */
import { describe, it } from "node:test";
import * as assert from "node:assert/strict";
import * as http from "node:http";
import { Memory } from "../memory.js";

async function withServer(
  handler: (request: http.IncomingMessage, response: http.ServerResponse, body: string) => void,
  run: (baseUrl: string) => Promise<void>,
): Promise<void> {
  const server = http.createServer((request, response) => {
    let body = "";
    request.on("data", (chunk) => { body += chunk; });
    request.on("end", () => handler(request, response, body));
  });
  await new Promise<void>((resolve) => server.listen(0, "127.0.0.1", resolve));
  const address = server.address() as { port: number };
  try {
    await run(`http://127.0.0.1:${address.port}`);
  } finally {
    await new Promise<void>((resolve) => server.close(() => resolve()));
  }
}

describe("score is not writable through update()", () => {
  it("update() throws instead of silently dropping score", async () => {
    let serverWasCalled = false;
    await withServer(
      (_request, response) => { serverWasCalled = true; response.end("{}"); },
      async (baseUrl) => {
        const memory = new Memory({ apiKey: "test-key", url: baseUrl });
        await assert.rejects(
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          () => memory.update(42, { score: 8 } as any),
          /setScore/,
          "a mensagem tem que nomear o método que funciona",
        );
        assert.equal(serverWasCalled, false,
          "a guarda deixou a requisição sair — tem que barrar antes do transporte");
      },
    );
  });

  it("update() throws even when score rides along other fields", async () => {
    await withServer(
      (_request, response) => response.end("{}"),
      async (baseUrl) => {
        const memory = new Memory({ apiKey: "test-key", url: baseUrl });
        await assert.rejects(
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          () => memory.update(42, { summary: "novo", score: 8 } as any),
          /setScore/,
          "sucesso parcial (summary grava, score não) é perda silenciosa",
        );
      },
    );
  });

  it("update() still accepts ordinary fields", async () => {
    let capturedPath = "";
    let capturedMethod = "";
    await withServer(
      (request, response) => {
        capturedPath = request.url ?? "";
        capturedMethod = request.method ?? "";
        response.end("{}");
      },
      async (baseUrl) => {
        const memory = new Memory({ apiKey: "test-key", url: baseUrl });
        await memory.update(42, { summary: "novo" });
        assert.equal(capturedMethod, "PATCH");
        assert.equal(capturedPath, "/api/v1/records/42");
      },
    );
  });

  it("setScore() posts to the durable replicated route with ids and score", async () => {
    let capturedPath = "";
    let capturedMethod = "";
    let capturedBody = "";
    await withServer(
      (request, response, body) => {
        capturedPath = request.url ?? "";
        capturedMethod = request.method ?? "";
        capturedBody = body;
        response.setHeader("Content-Type", "application/json");
        response.end(`{"status":"ok"}`);
      },
      async (baseUrl) => {
        const memory = new Memory({ apiKey: "test-key", url: baseUrl });
        await memory.setScore(42, 8);
        assert.equal(capturedMethod, "POST");
        assert.equal(capturedPath, "/api/v1/records/set-score");
        const parsed = JSON.parse(capturedBody);
        assert.deepEqual(parsed.ids, [42]);
        assert.equal(parsed.score, 8);
      },
    );
  });

  it("setScore() validates the schema range, boundaries included", async () => {
    await withServer(
      (_request, response) => {
        response.setHeader("Content-Type", "application/json");
        response.end(`{"status":"ok"}`);
      },
      async (baseUrl) => {
        const memory = new Memory({ apiKey: "test-key", url: baseUrl });
        for (const invalid of [0, -1, 11, 99, 2.5]) {
          await assert.rejects(() => memory.setScore(42, invalid), /between 1 and 10/);
        }
        // 1 e 10 são válidos: a guarda não pode encolher a faixa do schema.
        await memory.setScore(42, 1);
        await memory.setScore(42, 10);
      },
    );
  });
});
